Lib/mastermind.rb

```diff
@@ -1,24 +1,97 @@
-class Mastermind
-  def show_instructions
-    puts 'Welcome to Mastermind! Your challenge is
-to guess the secret four digit code in under 10 tries.
 
-The code can include the following colors:
+class Welcome
+  def show_instructions
+    puts "Welcome to Mastermind! Your challenge is
+to guess the secret four digit code in under 10 tries."
+    sleep 5
+    puts ""
+    puts "The code can include the following colors:
 Red(r), Green(g), Orange(o), Blue(b), Yellow(y) and Purple(p)
-Enter your guess by entering four color values from above, such as rgyb or pggy.
+Enter your guess by entering four color values from above, such as rgyb or pggy."
+    sleep 7
+    puts ""
+    puts "If you have a correct color in the right position, you get a red pin.
+If you have a correct color in the wrong position, you get a white pin"
+    sleep 5
+    puts ""
+    puts 'press \'q\' to quit, or hit enter to continue.'
+    quit if gets.chomp.downcase == "q"
+  end
 
-If you have a correct color in a correct position, you get a white pin.
-If you have a correct color in a different position, you get a black pin
-press \'q\' to quit, or enter your first guess.'
+  def quit
+    abort("It's been an honor serving with you. Farewell!")
   end
 end
 
 class Code
   attr_reader :answer
+  attr_reader :player_input
+  attr_reader :red_pins
+  attr_reader :white_pins
+  attr_reader :position
+  attr_reader :secret
+
 
   def generate_code
     @colors = colors = ["r", "g", "o", "y", "b", "p"]
-    answer = (1..4).map { @colors[rand(6)] }
+    @answer = (1..4).map { @colors[rand(6)] }
+  end
+
+  def red_counter(key, input)
+    red_pins = []
+    position = 0
+    while input.length > position
+      key.each do |match|
+        red_pins << 1 if key[position] == input[position]
+        position += 1
+      end
+    end
+    red_pins
+  end
+
+  def trim(key, input)
+    position = 0
+    if input[position] == key[position]
+      input[position] = ""
+      key[position] = ""
+    end
+    position += 1
+    if input[position] == key[position]
+      input[position] = ""
+      key[position] = ""
+    end
+    position += 1
+    if input[position] == key[position]
+      input[position] = ""
+      key[position] = ""
+    end
+    position += 1
+    if input[position] == key[position]
+      input[position] = ""
+      key[position] = ""
+    end
+    position += 1
+    input
+  end
+
+  def white_counter(key, input)
+    x = (key & input).flat_map { |n| [n]*[key.count(n), input.count(n)].min }
+    white_pins = x.reject { |c| c.empty? }
```

Comment:
You don't need to save `x.reject { |c| c.empty? }` as a local variable if you're not using it. You can just put the `reject` expression as the last line of the method.